Add a query throttle middleware

senior opened the pull request:
This will keep track of the number of concurrent queries going through
the query pipeline. By default it will allow half of the total Jetty
connections to be used for querying via the query pipeline. By default
we have 50 Jetty connections, so out of the box this number os
25. The number of query threads can be overridden vya
MB_MAX_CONCURRENT_QUERIES. Once we have reached the max number of
concurrent queries, subsequent queries will wait 5 seconds for a
slot, then return a 503.

iethree committed:
Add a query throttle middleware

This will keep track of the number of concurrent queries going through
the query pipeline. By default it will allow half of the total Jetty
connections to be used for querying via the query pipeline. By default
we have 50 Jetty connections, so out of the box this number os
25. The number of query threads can be overridden vya
MB_MAX_CONCURRENT_QUERIES. Once we have reached the max number of
concurrent queries, subsequent queries will wait 5 seconds for a
slot, then return a 503.

sbelak commented on src/metabase/query_processor/middleware/add_query_throttle.clj:
```diff
@@ -0,0 +1,48 @@
+            [puppetlabs.i18n.core :refer [tru]])
+  (:import [java.util.concurrent Semaphore TimeUnit]))
+
+(defn- calculate-max-queries-from-max-threads []
```
Semantic nitpicking, but why is this a defn, rather than a def?

senior replied:
Good catch, I'll fix that

sbelak reviewed (approved):
LGTM.

Two things that might be good to cover with tests in addition are query cancelation and streaming.

camsaul commented on src/metabase/query_processor/middleware/add_query_throttle.clj:
```diff
@@ -0,0 +1,48 @@
+(defn- throw-503-unavailable
+  []
+  (throw (ex-info (tru "Max concurrent query limit reached")
+           {:status 503})))
```
I think it's supposed to be `status-code`
But the QP middleware that catches exceptions automatically is going to catch this so we need to put some logic there too to make sure the exceptions come back the way we want

camsaul replied:
or maybe it's already working?

camsaul reviewed (changes requested):
I don't think this would work as-is because the catch-exception-middleware is going to catch the 503 exception and wrap it in the wacky 200 "query failure" response format. I think we need to tweak that middleware to let this pass thru

It would be good do a test where you combine the throttle middleware + catch exceptions middleware
and make sure it comes thru the way we want

senior commented:
@camsaul I agree. I'll dial that semaphore down to one and try running two queries through using the full stack.

camsaul commented:
@senior since we're only really worried about the interplay between those two middleware functions
you should just be able to test a query with something like this

```clj
(let [my-qp (->> identity
                 (#'throttle/throttle-queries my-semaphore)
                 catch-exceptions/catch-exceptions)]
  (my-qp my-query))..)
```

senior commented:
@camsaul looks like we repackage our query exceptions in two places. I found another spot [here](https://github.com/metabase/metabase/blob/master/src/metabase/query_processor.clj#L267). Any idea why we'd be doing that twice?

camsaul commented:
@senior I think that try-catch in `run-and-save-query!` is legacy and we can remove it since Exceptions are already handled by the middleware.

iethree committed:
Add an environment var to disable login/session throttling

senior commented:
@camsaul :confused: Looks like it must be used for some things as it breaks a bunch of tests if I remove it. Looks like the failures/errors go in two categories. The first is more specific error messages from embed_test, seeing `"Table \"XYZ\" not found"` when ` "An error occurred while running the query."` was expected. The errors look to be related to 500s coming back rather than 200s with our error payload. There's a pulse failure that just looks like a mistake but there's a `dataset_test` failure that's expecting a 200 with a failure payload rather than a 500.

 I'll write up a ticket for fixing that and make the change in both places for now.

iethree committed:
Fix SystemLocalizedString error while running FE tests

iethree committed:
Fix a docker/chome karma test config issue

When running tests in a docker container as root you'll get the
following error "Running as root without --no-sandbox is not
supported". Adding the `--no-sandbox` flag to the invocation fixes it.

iethree committed:
Add 503 retry logic

iethree committed:
Fix a Clojure linter issue

iethree committed:
Add a FlowFixMe to work around a type issue

iethree committed:
yarn.lock update

iethree committed:
Backend test fixes

iethree committed:
Minor FE unit test refactoring

tlrobinson reviewed (commented):
@metabase/core-developers Regarding the frontend changes, are we **sure** we want to default GET and especially POST requests to retry?

Currently it's only specifically for 503 errors. Is it always safe to retry on a 503 error?

Also if we're ok with retrying POSTs that return a 503 then why not PUT and DELETE?

tlrobinson committed:
api.js: merge options and retryOptions

tlrobinson committed:
Merge branch 'query-throttle' of github.com:metabase/metabase into query-throttle

tlrobinson commented on frontend/src/metabase/lib/api.js:
```diff
@@ -4,14 +4,40 @@ import querystring from "querystring";
+const MAX_RETRIES = 10;
+
+const DEFAULT_RETRY: RetryOptions = {
+  maxRetries: MAX_RETRIES,
```
The default of 10 retries + exponential backoff means it will take about 17 minutes for the request to finally fail. That seems long to me.

mazameli commented:
Tested this out for about 30 minutes with the instance generator tool going with the recommended settings, query against 4 DBs. Observations:

- Metabase didn't seem to slow down noticeably. My machine overall felt a bit slower, but I suspect that was from the instance-generator tool running locally. I ran this branch while using a snapshot of our app db, and did a whole host of things like x-rays, browsing questions, loading up and filtering dashboards, editing dashboards, opening saved questions, etc.
- Our "TV" dashboard seemed to load up in about the same time as on our stats instance; maybe even a little faster.
- I noticed some very slight but perceptible delays in rendering long lists of questions while scrolling, where the questions would pop in after a couple millisecond delay.

Let me know if I should test any other specific things.

iethree committed:
Changed query throttling to opt-in

This commit flips the logic for the query throttle. With this commit
query throttling is off by default. Users wanting to enable it will
use `MB_ENABLE_QUERY_THROTTLE=true`.
